### ISSUE TITLE
"1<<31" should not trigger a warning

### ISSUE THREAD

**Issue by dbabokin** [ispc/ispc#1826]
``1<<31`` should not trigger a warning. Even through, 1 is treated as ``int32``, left shift should not trigger a warning, as long as result is not exeding 32 bits. ISPC should follow C/C++ in this behavior.

Reproducer:
```c
uniform unsigned int32 foo() {
    uniform unsigned int32 u = 1<<31;
    return u;
}
```

```bash
>ispc test.ispc --target=avx2-i32x8
Warning: No output file or header file name specified. Program will be compiled and warnings/errors will be issued, but no output will be generated.
test.ispc:2:32: Warning: Binary expression with type "const uniform int32" can't represent value.
    uniform unsigned int32 u = 1<<31;
                               ^^^^^
```

**Comment by DeepakRajendrakumaran** [comment]
The issue here occurs during constant folding.
Currently ISPC uses it's implementation to optimize constant arithmetic.

When we see something like 1 << 20,  we have 1 and 20 which are int. But inorder to ensure that there is no data loss/overflow with the calculation, we do an additional check

We check if ((int)1 << (int)20) is equal to (long) 1 << (long)20. This check works for most normal cases except when shifting results in sign bit being set.

In case of 1 << 31,

((int)1 << (int)20) = 10000000000000000000000000000000 = -2147483648
((long)v0[i] << (long) v1[i]) = 0000000000000000000000000000000010000000000000000000000000000000 = 2147483648

This what causes the errant warning here

**Comment by dbabokin** [comment]
This is clearly not following C++ behavior when sign bit is affected. We need to count number of leading zeros in RHS, and require LHS to be less or equal to it.

**Comment by DeepakRajendrakumaran** [comment]
Possible solution:
https://github.com/DeepakRajendrakumaran/ispc/commits/1826

I think this is a much better check. Here I’m checking if the number of bits to be shifter is greater than or equal to the width of the value to be shifted.

**Comment by dbabokin** [comment]
No, you check different thing. You are checking if shift count is larger than type size. But you really need to look at the number of leading zeroes in the value.
I.e. 0x7fffffff has 1 leading zero. It's ok to shift left by 1, but if shifting by more than 1, this need to trigger a warning.

Also you've forgot a ``break;`` statement in the preceding cases.

**Comment by DeepakRajendrakumaran** [comment]
Ah...nevermind. I had a brain freeze. Yep, will change it.

**Comment by dbabokin** [comment]
Fixed by #1866
